### PR TITLE
volume-migration: avoid to start a vm on failed migration

### PR DESCRIPTION
**Pull request by alicefr**
### What this PR does
Before this PR:
Possible to start a VM after a failed volume migration with an inconsistent volume set. In this case, the VM could be started from an unbootable disk

After this PR:
Avoid to start a VM if the ManualRecoveryRequiered is set.

### Release note

```release-note
NONE
```

### CODE REVIEW

**Comment by alicefr** [comment]
/hold
It requires https://github.com/kubevirt/kubevirt/pull/12355 to be merged

**Comment by fossedihelm** [comment]
@alicefr Are you able to rebase this on top of the previous one?
As @vladikr pointed out, we should have this in the same release. Thanks

File: pkg/virt-api/rest/subresource.go [hunk 1]
```diff
@@ -472,6 +472,13 @@ func (app *SubresourceAPIApp) StartVMRequestHandler(request *restful.Request, re
 		writeError(errors.NewConflict(v1.Resource("virtualmachine"), name, fmt.Errorf("VM is already running")), response)
 		return
 	}
+	if vm.Status.VolumeUpdateState != nil &&
+		vm.Status.VolumeUpdateState.VolumeMigrationState != nil &&
+		vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired != nil &&
+		*vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired {
+		writeError(errors.NewConflict(v1.Resource("virtualmachine"), name, fmt.Errorf("VM needs to be manually recovered because the volume migration failed and the volumes are in an incosistent state")), response)
```
Comment (vladikr) [inline pkg/virt-api/rest/subresource.go:479]: maybe something like "VM recovery required: Volume migration failed, leaving some volumes pointing to non-existent targets; manual intervention is needed to reassign them to their original volumes." ?

File: pkg/virt-controller/watch/vm/vm.go [hunk 2]
```diff
@@ -1621,10 +1621,32 @@ func syncVolumeMigration(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineIn
 	}
 	vmCond := controller.NewVirtualMachineConditionManager()
 	vmiCond := controller.NewVirtualMachineInstanceConditionManager()
-	// Something went wrong with the VMI while the volume migration was in progress
-	if vmi == nil && vmCond.HasConditionWithStatus(vm, virtv1.VirtualMachineConditionType(virtv1.VirtualMachineInstanceVolumesChange), k8score.ConditionTrue) {
+
+	srcMigVols := make(map[string]string)
+	for _, v := range vm.Status.VolumeUpdateState.VolumeMigrationState.MigratedVolumes {
```
Comment (vladikr) [inline pkg/virt-controller/watch/vm/vm.go:1626]: maybe it needs a simple comment, "// Checks if VM volumes have been recovered to their original PVCs after a failed migration. "

**Review by vladikr — COMMENTED** [review]
The PR looks solid to me. Thank @alicefr.
Just a few nits. 

It begs a question though - if we already know the source PVCs, can't we recover the volumes automatically?

File: pkg/virt-controller/watch/vm/vm.go [hunk 3]
```diff
@@ -1621,10 +1621,32 @@ func syncVolumeMigration(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineIn
...
+			continue
+		}
+		if origName != name {
+			recoveredOldVMVolumes = false
+		}
+	}
+	if recoveredOldVMVolumes {
+		vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired = nil
+		return
+	} else if vmi == nil && vmCond.HasConditionWithStatus(vm, virtv1.VirtualMachineConditionType(virtv1.VirtualMachineInstanceVolumesChange), k8score.ConditionTrue) {
```
Comment (fossedihelm) [inline pkg/virt-controller/watch/vm/vm.go:1645]: nit: the else can be dropped
```suggestion
	}

	if vmi == nil && vmCond.HasConditionWithStatus(vm, virtv1.VirtualMachineConditionType(virtv1.VirtualMachineInstanceVolumesChange), k8score.ConditionTrue) {
		// Something went wrong with the VMI while the volume migration was in progress
		vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired = pointer.P(true)
	}
```

**Review by fossedihelm — COMMENTED** [review]
Thank you @alicefr!
Looks good overall. Apart from the small nit below, I have two things:
1. Shouldn't we consider the restart too? I mean, if a VM is restarted during the migration, shouldn't we remain in an inconsistent state?
2. Maybe we are missing a unit test in `subresource_test`

Thanks

**Comment by alicefr** [comment]
> It begs a question though - if we already know the source PVCs, can't we recover the volumes automatically?

No, because it will be a violation of the declarative approach we have been following for this feature. In order to recover the volumes automatically, we would need to patch the volume spec in behalf of the users.

**Comment by alicefr** [comment]
> 2. Maybe we are missing a unit test in `subresource_test`
@fossedihelm Done, PTAL

**Comment by vladikr** [comment]
> > It begs a question though - if we already know the source PVCs, can't we recover the volumes automatically?
> 
> No, because it will be a violation of the declarative approach we have been following for this feature. In order to recover the volumes automatically, we would need to patch the volume in the VM spec in behalf of the users.
> 
> EDIT: this should be done by the overlaying tools for the migration plans

I meant that we would boot from the known source volume from the status and then the migration would be re-triggered.
This wouldn't violate the declarative approach but would improve the user experience.

However, I think this can be done in the future.
Currently, I'm more concerned about VM potentially reaching an unbootable state.

/approve

File: pkg/virt-controller/watch/vm/vm.go [hunk 4]
```diff
@@ -1621,10 +1621,34 @@ func syncVolumeMigration(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineIn
...
+		origName, ok := srcMigVols[v.Name]
+		if !ok {
+			continue
+		}
+		if origName != name {
+			recoveredOldVMVolumes = false
+		}
+	}
+	if recoveredOldVMVolumes {
+		vm.Status.VolumeUpdateState.VolumeMigrationState.ManualRecoveryRequired = nil
```
Comment (mhenriks) [inline pkg/virt-controller/watch/vm/vm.go:1644]: Unit test for this?

Comment (alicefr) [reply pkg/virt-controller/watch/vm/vm.go:1644]: I need to add entirely for the sycVolumeMigration function? Can we do it as a follow back?

**Comment by alicefr** [comment]
@fossedihelm the issue is that the test isn't added by this PR but it is already upstream. Then, we have a flake :/

File: pkg/virt-api/rest/subresource_test.go [hunk 5]
```diff
@@ -41,10 +41,12 @@ import (
 	"github.com/onsi/gomega/ghttp"
 
 	k8sv1 "k8s.io/api/core/v1"
+	v12 "k8s.io/api/core/v1"
```
Comment (fossedihelm) [inline pkg/virt-api/rest/subresource_test.go:44]: please use the already present `k8sv1`

File: tests/storage/migration.go [hunk 6]
```diff
@@ -54,6 +54,7 @@ import (
 	"kubevirt.io/kubevirt/tests/framework/checks"
 	"kubevirt.io/kubevirt/tests/framework/kubevirt"
 	"kubevirt.io/kubevirt/tests/framework/matcher"
+	. "kubevirt.io/kubevirt/tests/framework/matcher"
```
Comment (fossedihelm) [inline tests/storage/migration.go:57]: duplicate import`

File: tests/storage/migration.go [hunk 7]
```diff
@@ -539,6 +539,23 @@ var _ = SIGDescribe("[Serial]Volumes update with migration", Serial, func() {
...
+			By("Update volumes")
+			updateVMWithPVC(vm, volName, destPVC)
+			waitMigrationToExist(vm.Name, ns)
+			waitVMIToHaveVolumeChangeCond(vm.Name, ns)
+
+			By("Restarting the VM during the volume migration")
+			restartOptions := &virtv1.RestartOptions{GracePeriodSeconds: pointer.P(int64(0))}
+			err := virtClient.VirtualMachine(vm.Namespace).Restart(context.Background(), vm.Name, restartOptions)
+			Expect(err).To(MatchError(ContainSubstring("VM recovery required")))
+		})
```
Comment (fossedihelm) [inline tests/storage/migration.go:557]: Can't this be merged with the other one for start?
I mean, first try to restart, ensure that the request is blocked, then stop and try to start

**Review by fossedihelm — COMMENTED** [review]
Thank you! 
Some small comments below

File: pkg/virt-controller/watch/vm/vm.go (hunk 4 above)
Comment (mhenriks) [reply pkg/virt-controller/watch/vm/vm.go]: I don't see any coverage in unittests for `status.volumeUpdateState` but can be done in follow up

Comment (alicefr) [reply pkg/virt-controller/watch/vm/vm.go]: I added the unit tests please, check the last commit

**Comment by alicefr** [comment]
I fixed the last comments and added also the unit tests for the `syncVolumeMigration`. In order to properly write the unit tests, I have added the `WithVMStatus` function.
@mhenriks @fossedihelm PTAL

@EdDev can you please take a look to commit 3a6460b58477dbec0964d8a2a6badc24bc9c9582 , it was required to write the status of the VM in the unit tests.

File: pkg/libvmi/status/status.go [hunk 8]
```diff
@@ -133,3 +133,35 @@ func WithInterfaceStatus(interfaceStatus v1.VirtualMachineInstanceNetworkInterfa
...
+func NewVMStatus(opts ...VMOption) v1.VirtualMachineStatus {
+	vmStatus := &v1.VirtualMachineStatus{}
+	for _, f := range opts {
+		f(vmStatus)
+	}
+
+	return *vmStatus
+}
+
+func WithVolumeUpdateState(volumeUpdateState *v1.VolumeUpdateState) VMOption {
```
Comment (fossedihelm) [inline pkg/libvmi/status/status.go:157]: Currently it is not present but in future we could be add the `WithVolumeUpdateState` for vmi.
I would stay with using `VM` for VM related options
```suggestion
func WithVMVolumeUpdateState(volumeUpdateState *v1.VolumeUpdateState) VMOption {
```

**Comment by mhenriks** [comment]
/lgtm
/hold

holding for @fossedihelm and @EdDev

**Comment by fossedihelm** [comment]
/lgtm
Thank you @alicefr 
@EdDev up to you to unhold

**Comment by alicefr** [comment]
/unhold
Got good feedback from @EdDev

**Comment by fossedihelm** [comment]
/hold
@alicefr can you have a look at the failures?

**Pull request merged**